<div class="text-cell-section-header">
<h1 style="text-align: center;"><strong>PROGRAMACI&Oacute;N PARA EL AN&Aacute;LISIS DE DATOS</strong></h1>
</div>
<h1 style="text-align: center;"><strong>Tarea 2
Desarrollo de código para el tratamiento y manejo de
información</strong></h1>
<h1 style="text-align: center;"><strong>PRESENTADO POR: DARWIN MERCADO</strong></h1>
<h1 style="text-align: center;"><strong>GRUPO 203008069_3</strong></h1>
<h1 style="text-align: center;"><strong>PRESENTADO A: FELIPE ALEXANDER PIPICANO GUZMAN</strong></h1>
<h1 style="text-align: center;"><strong>Universidad Nacional Abierta y a Distancia &ndash; UNAD</strong></h1>
<h1 style="text-align: center;"><strong>22 de octubre de 2023</strong></h1>

# **PRESENTACION DEL EJERCICIO**

A continuación se va a desarrollar el ejercicio número 3 de la guía de actividad propuesta por el tutor, en este se va a trabajar con un DataFrame datos de pesos de estudiantes, se desarrollarán calculos de datos especificos, filtros de DataFrame y se relacionarán como trabajar con los Inner Join

# REVISIÓN TEÓRICA

- **Combinación de DataFrames:**
Esta función permite combinar dos o más DataFrames en 1 solo, usando las columnas de interés como los valores claves a analizar, donde se pueden realizar operaciones y toma de datos especificos.

- **Unión de DataFrames:**
Esta función permite combinar dos DataFrames a partir de los indices, lo cual arroja los datos de interés que se busquen entre los dos DataFrames, esto se puede hacer con la función $.join()$ o $.merge()$.

- **Concatenación de DataFrames:**
La concatenación se utiliza para colocar un DataFrame uno encima de otro o también a el lado, teniendo como función principal $.concat()$

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
df=pd.read_csv("Anexo 1 - Student Weight Status.csv", sep=",",header = 0)
df

LOCATION CODE       COUNTY         AREA NAME            REGION  \
0                 4  CATTARAUGUS       CATTARAUGUS  WESTERN NEW YORK   
1                 4  CATTARAUGUS       CATTARAUGUS  WESTERN NEW YORK   
2                 4  CATTARAUGUS       CATTARAUGUS  WESTERN NEW YORK   
3                 4  CATTARAUGUS       CATTARAUGUS  WESTERN NEW YORK   
4                 4  CATTARAUGUS       CATTARAUGUS  WESTERN NEW YORK   
...             ...          ...               ...               ...   
8917         700000          NaN  WESTERN NEW YORK  WESTERN NEW YORK   
8918         700000          NaN  WESTERN NEW YORK  WESTERN NEW YORK   
8919         700000          NaN  WESTERN NEW YORK  WESTERN NEW YORK   
8920         700000          NaN  WESTERN NEW YORK  WESTERN NEW YORK   
8921         700000          NaN  WESTERN NEW YORK  WESTERN NEW YORK   

     SCHOOL YEARS  NO. OVERWEIGHT  PCT OVERWEIGHT  NO. OBESE  PCT OBESE  \
0       2010-2012           575.0            18.4      565.0       18.1   
1       2010-2012           380.0            19.4      416.0       21.3   
2       2012-2014           873.0            17.2      994.0       19.6   
3       2012-2014           551.0            17.2      571.0       17.8   
4       2012-2014           292.0            16.0      413.0       22.6   
...           ...             ...             ...        ...        ...   
8917    2014-2016          6080.0            16.3     5712.0       15.3   
8918    2014-2016          4495.0            17.8     4958.0       19.7   
8919    2016-2018         10422.0            16.3    11206.0       17.5   
8920    2016-2018          6089.0            15.7     6118.0       15.8   
8921    2016-2018          4224.0            17.2     5061.0       20.7   

      NO. OVERWEIGHT OR OBESE  PCT OVERWEIGHT OR OBESE     GRADE LEVEL  \
0                      1140.0                     36.5      ELEMENTARY   
1                       796.0                     40.7     MIDDLE/HIGH   
2                      1867.0                     36.9  DISTRICT TOTAL   
3                      1122.0                     35.0      ELEMENTARY   
4                       705.0                     38.6     MIDDLE/HIGH   
...                       ...                      ...             ...   
8917                  11792.0                     31.6      ELEMENTARY   
8918                   9453.0                     37.6     MIDDLE/HIGH   
8919                  21628.0                     33.8  DISTRICT TOTAL   
8920                  12207.0                     31.6      ELEMENTARY   
8921                   9285.0                     37.9     MIDDLE/HIGH   

     AREA TYPE STREET ADDRESS CITY STATE  ZIP CODE  LATITUDE  LONGITUDE  \
0       COUNTY            NaN  NaN   NaN       NaN       NaN        NaN   
1       COUNTY            NaN  NaN   NaN       NaN       NaN        NaN   
2       COUNTY            NaN  NaN   NaN       NaN       NaN        NaN   
3       COUNTY            NaN  NaN   NaN       NaN       NaN        NaN   
4       COUNTY            NaN  NaN   NaN       NaN       NaN        NaN   
...        ...            ...  ...   ...       ...       ...        ...   
8917    REGION            NaN  NaN   NaN       NaN       NaN        NaN   
8918    REGION            NaN  NaN   NaN       NaN       NaN        NaN   
8919    REGION            NaN  NaN   NaN       NaN       NaN        NaN   
8920    REGION            NaN  NaN   NaN       NaN       NaN        NaN   
8921    REGION            NaN  NaN   NaN       NaN       NaN        NaN   

     Location 1  
0           NaN  
1           NaN  
2           NaN  
3           NaN  
4           NaN  
...         ...  
8917        NaN  
8918        NaN  
8919        NaN  
8920        NaN  
8921        NaN  

[8922 rows x 20 columns]

**- Cree un DataFrame p1 con los datos del 0 al 1 con las columnas
'area_name' y 'county' de df.**

In [ ]:
p1= df[["AREA NAME","COUNTY"]][0:2]

AREA NAME       COUNTY
0  CATTARAUGUS  CATTARAUGUS
1  CATTARAUGUS  CATTARAUGUS

- ***Para poder crear el DataFrame p1 y obtener los datos del 0 al 1 debimos de tomar los datos desde el cero a el 2 de las columnas AREA NAME y COUNTY, adicional a esto se puede utilizar la función .loc y esta nos da la ubicación exacta sin necesidad de ir hasta el número 2***

**- Creen un DataFrame p2 con los datos del 2 al 4 con las
columnas 'area_name' y 'county' de df.**


In [ ]:
p2= df[["AREA NAME","COUNTY"]][2:5]
p2

AREA NAME       COUNTY
2  CATTARAUGUS  CATTARAUGUS
3  CATTARAUGUS  CATTARAUGUS
4  CATTARAUGUS  CATTARAUGUS

- ***Para poder realizar el DataFrame p2 y obtener los datos del 2 al 4 se utiliza el mismo metodo que en el punto anterior, empezando desde el número 2 hasta la posición 5 para obtener las posiciones desde el 2 al 4 y se filtra el DataFrame inicial por AREA NAME y COUNTY***

**- Concatene p1 y p2 y muestre los resultados.**

In [ ]:
pd.concat([p1,p2])

AREA NAME       COUNTY
0  CATTARAUGUS  CATTARAUGUS
1  CATTARAUGUS  CATTARAUGUS
2  CATTARAUGUS  CATTARAUGUS
3  CATTARAUGUS  CATTARAUGUS
4  CATTARAUGUS  CATTARAUGUS

- ***Usamos la función concat para unir p1 y p2 que fueron creados anteriormente y poder visualizar todos los datos combinados***

**- Concatene p1 y p2 identificando las asignaciones por las llaves
‘p1’ y ‘p2’ y muestre los resultados y asígnelos a la variable
concatened. Comente las diferencias con el DataFrame anterior
y mencione desde su punto de vista las posibles utilidades de
esta herramienta.**


In [ ]:
concatenated = pd.concat([p1,p2], keys={"p1","p2"})
print(concatenated)

        AREA NAME       COUNTY
p1 0  CATTARAUGUS  CATTARAUGUS
   1  CATTARAUGUS  CATTARAUGUS
p2 2  CATTARAUGUS  CATTARAUGUS
   3  CATTARAUGUS  CATTARAUGUS
   4  CATTARAUGUS  CATTARAUGUS


- ***En este nuevo DataFrame lo que tenemos es que hay unas llaves especificas para cada valor de p1 y p2, a diferencia del pasado donde los datos erán simplemente un DataFrame común.***
- ***Esta herramienta sería muy útil para diferencia ciertos datos que sean parecidos en un DataFrame que tenga datos por ejemplo de notas de colegio que sean parecidas y con esto podemos diferencia a cual estudiante pertenece cada nota y no tener complicaciones***

**- Desde el DataFrame concatened rescate los datos de p1.**

In [ ]:
concatenated.loc["p1"]

AREA NAME       COUNTY
0  CATTARAUGUS  CATTARAUGUS
1  CATTARAUGUS  CATTARAUGUS

- ***Se pudo rescatar los datos de p1 a través de la función .loc que nos arroja la posición y valores exactos de p1***

#**Operaciones de agregación:**

**- Encuentre el número promedio de estudiantes que son obesos
en escuelas de primaria (tipo ‘ELEMENTARY’).**

In [ ]:
prom = df[["GRADE LEVEL","NO. OBESE"]] ## Creamos un DF nuevo donde vamos a tener las variables de interes

In [ ]:
promedio =prom[prom["GRADE LEVEL"] == "ELEMENTARY"] ## se realiza el filtro, donde buscamos que el grado sea primaria
promedio.mean(numeric_only=True)## Usamos la función .mean indicandole que solo tome los valores númericos y así evitar tener errores

NO. OBESE    275.786762
dtype: float64

- ***Se puede observar que el promero de estudiantes obesos en las escuelas de primaria es de 275.786, lo cual es un número bastante elevado para ser solo primaria, se deberían de plantear cuidados más agresivos para este tipo de estudiantes***

**- Encuentre el número total de estudiantes de primaria que son
obesos.**

In [ ]:
df[df["GRADE LEVEL"] == "ELEMENTARY"]["NO. OBESE"].sum() ## Realizamos un filtro donde decimos que sean solo de primaria y que sean obesos, luego sumamos la cantidad con .sum()

770824.0

- ***Se tiene como número total de obesos en primaria 770824 que se obtuvo utilizando el filtro por primaria en GRADE LEVEL sobre el total de obesos y luego usar la función .sum()para sumar los totales. Sigue siendo un dato preocupante, ya que es un valor elevado para ser grado de primaria***

**- Encuentre el número máximo de estudiantes obsesos que tiene
una escuela.**

In [ ]:
df["NO. OBESE"].max() ## Calculamos el valor maximo con .max() para obtener la escuela con el valor maximo de estudiantes obesos

86016.0

***El resultado fue de 86016 es el número máximo de obesos que tiene una escuela, lo cual es un valor bastante predominante sabiendo que son adolescentes, jovenes y niños***

**- Encuentre el número mínimo de estudiantes obsesos que tiene
una escuela.**

In [ ]:
df["NO. OBESE"].min() ## calculamos el valor minimo de obesos que tienen las escuelas con .min()

5.0

***El resultado de 5 nos da un valor bastante bajo a comparación de los otros datos obtenidos, teniendo en cuenta esto las demas escuelas deberían de optar por seguir los caminos de cuidado de esta escuela en particular***

**- Obtenga la desviación estándar del número de estudiantes
obsesos en las escuelas.**

In [ ]:
df["NO. OBESE"].std() ## obtenemos la desviación estándar con .std() de los estudiantes obesos

2519.8439356637937

***El total obtenido de 2519.8439356637937 nos da nota de un número bastante elevado entre cada escuela, por lo tanto es de preocupación.***

In [ ]:
Primary = df[(df["COUNTY"] == "DELAWARE") & (df["GRADE LEVEL"] == "ELEMENTARY")] ## Se indica un filtro donde la variable Primary tome los datos de COUNTY del condado DELAWARE y de GRADE LEVEL indicamos que sean ELEMENTARY del df inicial
delaware_primay = Primary.shape[0]
print(delaware_primay)

52


***Se tiene que la cantidad de escuelas primarias en el condado de DELAWARE es de 52***

# **UNIONES (JOINS):**

**- Defina un DataFrame de búsqueda llamado grade_lookup, que
tenga las columnas ‘grade_level y ‘level’ que asigne niveles a
cada uno de los grados como sigue: 1 a 'ELEMENTARY', 2 a
'MIDDLE/HIGH', 3 a 'MISC'.**

In [ ]:
datos= {"GRADE LEVEL" : ["ELEMENTARY", "MIDDLE/HIGH", "MISC"], "level" :[1,2,3] }
grade_lookup=pd.DataFrame(datos)
grade_lookup

GRADE LEVEL  level
0   ELEMENTARY      1
1  MIDDLE/HIGH      2
2         MISC      3

***- Se definieron los keys values de GRADE LEVEL Y level que son las variables de interés con las que vamos a trabajar, luego se definieron como un DataFrame, para poder trabajar sobre los datos mas adelante***

**- Por otro lado, tome las primeras cinco filas de la columna de
datos en la columna GRADE e imprímalas (Téngalas en cuenta
pues las utilizaremos para estudiar los tipos de uniones).**

In [ ]:
prueba = pd.DataFrame(df["GRADE LEVEL"].loc[0:4])

***- Se utiliza la función .loc para tomar los variables de las primeras 5 columnas y se muestran***

**- Describa los inner joins y realice un ejemplo con las primeras 5
filas de la columna grade_level y grade_lookup, muestre sus
resultados y explíquelos.**

- **Inner joins**: Se utilizan para encontrar un valor en especifico que tienen las dos tablas o DataFrames propuestos y se debe de realizar con la función .join() especificando que es será how = "inner".

In [ ]:
inn_join = prueba.join(grade_lookup.set_index(["GRADE LEVEL"]), on="GRADE LEVEL", how="inner")
inn_join

GRADE LEVEL  level
0   ELEMENTARY      1
3   ELEMENTARY      1
1  MIDDLE/HIGH      2
4  MIDDLE/HIGH      2

***Como se puede observar en la tabla obtenida, tenemos los datos que tienen en común las dos tablas, teniendo como función .join() se defininen las tablas que se van a usar y el metodo que en este caso es how = "inner"***

**- Describa los left outer joins y realice un ejemplo con las
primeras 5 filas de la columna grade_level y grade_lookup,
muestre sus resultados y explíquelos.**

- **Los left outer joins nos permiten determinar los valores en especifico de esa tabla y verificar con otra que referente hay en ambas partes, teniendo como resultado una tabla con todas las variables de la tabla de estudio.**

In [ ]:
inn_joinleft = prueba.join(grade_lookup.set_index(["GRADE LEVEL"]), on="GRADE LEVEL", how="left")
inn_joinleft

GRADE LEVEL  level
0      ELEMENTARY    1.0
1     MIDDLE/HIGH    2.0
2  DISTRICT TOTAL    NaN
3      ELEMENTARY    1.0
4     MIDDLE/HIGH    2.0

***Se utilizo la misma función .join() y se tomo como how = "left" ya que era un left join, donde analizamos los datos completos de la tabla GRADE LEVEL y la relación de grade_lookup***

**- Describa los full outer joins y realice un ejemplo con las
primeras 5 filas de la columna grade_level y grade_lookup,
muestre sus resultados y explíquelos.**

- **En los full outer joins se utiliza los datos de ambas tablas en común, sin tener en cuenta sus relaciones entre si, solamente los datos individuales de cada uno, teniendo como resultado el total de valores que contiene cada tabla en una sola.**

In [ ]:
inn_joinouter = prueba.join(grade_lookup.set_index(["GRADE LEVEL"]), on="GRADE LEVEL", how="outer")
inn_joinouter[["GRADE LEVEL", "level"]]

GRADE LEVEL  level
0.0      ELEMENTARY    1.0
3.0      ELEMENTARY    1.0
1.0     MIDDLE/HIGH    2.0
4.0     MIDDLE/HIGH    2.0
2.0  DISTRICT TOTAL    NaN
NaN            MISC    3.0

***En esta función se volvío a utilizar .join() junto con how = "outer" que nos indica que se hará un full outer join, arrojandonos los datos de ambas tablas sin necesidad de que haya una relación entre si.***

#**AGRUPACIONES**

- Encontrar la suma del número de estudiantes obesos en cada
uno de los grados mediante el uso de la función groupby de
pandas. Muestre sus resultados y explíquelo.

In [ ]:
from pandas.core import groupby
df["NO. OBESE"].groupby(df["GRADE LEVEL"]).sum()

GRADE LEVEL
DISTRICT TOTAL    1359576.0
ELEMENTARY         770824.0
MIDDLE/HIGH        577812.0
Name: NO. OBESE, dtype: float64

***En esta función pudimos encontrar la suma de todos los grados para los esutidantes obesos, pudimos observar que en la variable DISTRIC TOTAL hay un mayor número de estudiantes obesos.***

- Muestre diferentes tipos de agregación por suma, media y
desviación estándar de los estudiantes obsesos en cada uno de
los grados.

In [ ]:
df.groupby("GRADE LEVEL")["NO. OBESE"].agg(["sum","mean","std"])

sum        mean          std
GRADE LEVEL                                       
DISTRICT TOTAL  1359576.0  473.719861  3502.973168
ELEMENTARY       770824.0  275.786762  2021.124815
MIDDLE/HIGH      577812.0  215.039821  1532.055998

***Se realiza una función donde agregamos la suma, la media y la desviación estandar a partir de la agrupación por GRADE LEVEL en las personas obesas, donde encontramos que DISTRIC TOTAL en todas las relaciones estadisticas sigue teniendo mayor número***

#**CONCLUCIONES**

- Se pudo observar que en las variables que nos dieron para el desarrollo de este ejercicio, que hay una preocupación muy grande por los elevados casos de diabetes en los diferentes grados escolares, siendo con más impacto el valor de las escuelas primarias de casí 100.000 casos de obesidad de diferencia con Middle y high.

- Se pudo también aprender a como realizar operaciones entre DataFrames, como hacer busquedas con filtros de los valores de importancia, adicionar o quitar valores, como definir la localización especifica de un dato de importancia.

- También se utilizaron los Inner Joins que nos permiten buscar relaciones entre tablas, relaciones entre datos y otros.

# **REFERENCIAS**

- *Samir Madhavan. (2015). Mastering Python for Data Science: Explore the World of Data Science Through Python and Learn How to Make Sense of Data. Packt Publishing. (pp. 11-25). https://bibliotecavirtual.unad.edu.co/login?url=https://search.ebscohost.com/login.aspx?direct=true&db=nlebk&AN=1058787&lang=es&site=ehost-live&ebv=EB&ppid=pp_11*